tutorial/second.nim

```diff
@@ -0,0 +1,151 @@
+when not(compileOption("threads")):
+  {.fatal: "Please, compile this program with the --threads:on option!".}
+
+import tables, strformat, strutils
+import chronos                              
+import ../libp2p/[switch,                   
+                  multistream,              
+                  crypto/crypto,            
+                  protocols/identify,       
+                  connection,               
+                  transports/transport,     
+                  transports/tcptransport,  
+                  multiaddress,             
+                  peerinfo,                 
+                  peer,                     
+                  protocols/protocol,       
+                  protocols/secure/secure,  
+                  protocols/secure/secio,   
+                  muxers/muxer,            
+                  muxers/mplex/mplex,       
+                  muxers/mplex/types]     
+
+const ChatCodec = "/nim-libp2p/chat/1.0.0"
+const DefaultAddr = "/ip4/127.0.0.1/tcp/55505"
+
+const Help = """
+  Commands: /[?|hep|connect|disconnect|exit]
+  help: Prints this help
+  connect: dials a remote peer
+  disconnect: ends current session
+  exit: closes the chat
+"""
+
+type ChatProto = ref object of LPProtocol
+  switch: Switch          # a single entry point for dialing and listening to peer
+  transp: StreamTransport # transport streams between read & write file descriptor 
+  conn: Connection        # create and close read & write stream
+  connected: bool         # if the node is connected to another peer
+  started: bool           # if the node has started
+
+# copied from https://github.com/status-im/nim-beacon-chain/blob/0ed657e953740a92458f23033d47483ffa17ccb0/beacon_chain/eth2_network.nim#L109-L115 
+proc initAddress(T: type MultiAddress, str: string): T =
+  let address = MultiAddress.init(str)
+  if IPFS.match(address) and matchPartial(multiaddress.TCP, address):
+    result = address
+  else:
+    raise newException(MultiAddressError,
+                         "Invalid bootstrap node multi-address")
+
+proc dialPeer(p: ChatProto, address: string) {.async.} =
+  let multiAddr = MultiAddress.initAddress(address);
+  let parts = address.split("/")
+  let remotePeer = PeerInfo.init(parts[^1],
+                                 [multiAddr])
+
+  echo &"dialing peer: {multiAddr}"
+  p.conn = await p.switch.dial(remotePeer, ChatCodec)
+  p.connected = true
+
+proc readAndPrint(p: ChatProto) {.async.} =
+  while true:
+    while p.connected:
+      echo cast[string](await p.conn.readLp())
```

Comment:
let's use `string.fromBytes` here and elsewhere instead of casts (from stew/byteutils)

Reply:
hmm why not use cast?

Reply:
`cast` disables the type checking in the language and is not guaranteed to work across nim versions or implementations - it's a, unsafe low-level tool of last resort: https://nim-lang.org/docs/manual.html#statements-and-expressions-type-casts